File: android/src/main/java/com/supermap/smNative/SMSymbol.java
package com.supermap.smNative;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableArray;
import com.facebook.react.bridge.WritableMap;
import com.supermap.RNUtils.DataUtil;
import com.supermap.data.Resources;
import com.supermap.data.Symbol;
import com.supermap.data.SymbolFillLibrary;
import com.supermap.data.SymbolGroup;
import com.supermap.data.SymbolGroups;
import com.supermap.data.SymbolLibrary;
import com.supermap.data.SymbolLineLibrary;
import com.supermap.data.SymbolMarkerLibrary;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

public class SMSymbol {

    public static WritableArray getSymbolGroups(Resources resources, String type, String path) {
        SymbolLibrary symbolLibrary;
        WritableArray groupArr = Arguments.createArray();
        SymbolGroup group;

        if (type.equals("") && path.equals("")) {
            SymbolGroup rootGroup = resources.getFillLibrary().getRootGroup();
            rootGroup = resources.getMarkerLibrary().getRootGroup();
            WritableMap markerGroup = Arguments.createMap();
            markerGroup.putString("name", "点符号库");
            markerGroup.putInt("count", rootGroup.getCount());
            markerGroup.putArray("childGroups", findAllSymbolGroups(rootGroup, "marker", path));
            markerGroup.putString("path", rootGroup.getName());
            markerGroup.putString("type", "marker");
            groupArr.pushMap(markerGroup);

            rootGroup = resources.getLineLibrary().getRootGroup();
            WritableMap lineGroup = Arguments.createMap();
            lineGroup.putString("name", "线型符号库");
            lineGroup.putInt("count", rootGroup.getCount());
            lineGroup.putArray("childGroups", findAllSymbolGroups(rootGroup, "line", path));
            lineGroup.putString("path", rootGroup.getName());
            lineGroup.putString("type", "line");
            groupArr.pushMap(lineGroup);

            WritableMap fillGroup = Arguments.createMap();
            fillGroup.putString("name", "填充符号库");
            fillGroup.putInt("count", rootGroup.getCount());
            fillGroup.putArray("childGroups", findAllSymbolGroups(rootGroup, "fill", path));
            fillGroup.putString("path", rootGroup.getName());
            fillGroup.putString("type", "fill");
            groupArr.pushMap(fillGroup);
        } else if (type.equals("") && !path.equals("")) {
            groupArr.pushArray(findAllSymbolGroups(resources.getMarkerLibrary().getRootGroup(), type, path));
            groupArr.pushArray(findAllSymbolGroups(resources.getLineLibrary().getRootGroup(), type, path));
            groupArr.pushArray(findAllSymbolGroups(resources.getFillLibrary().getRootGroup(), type, path));
        } else if (!type.equals("")) {
            if (type.equals("fill")) {
                symbolLibrary = resources.getFillLibrary();
            } else if (type.equals("line")) {
                symbolLibrary = resources.getLineLibrary();
            } else {
                symbolLibrary = resources.getMarkerLibrary();
            }
            String _path = path.equals("") ? symbolLibrary.getRootGroup().getName() : path;

            if (path.equals("")) {
                group = symbolLibrary.getRootGroup();
                _path = group.getName();
            } else {
                group = findSymbolGroups(resources, type, _path);
                _path = group.getName();
            }

            WritableMap groupInfo = Arguments.createMap();
            groupInfo.putString("name", group.getName());
            groupInfo.putInt("count", group.getCount());
            groupInfo.putArray("childGroups", findAllSymbolGroups(group, type, _path));
            groupInfo.putString("path", _path);
            groupInfo.putString("type", type);
            groupArr.pushMap(groupInfo);
        }

        return groupArr;
    }

    public static WritableArray findAllSymbolGroups(SymbolGroup symbolGroup, String type, String path) {
        WritableArray groupArr = Arguments.createArray();
        SymbolGroups groups = symbolGroup.getChildGroups();

        for (int i = 0; i < groups.getCount(); i++) {
            SymbolGroup group = groups.get(i);
            WritableArray childGroupArr = Arguments.createArray();
            if (group.getChildGroups().getCount() > 0) {
                childGroupArr = findAllSymbolGroups(group, type, path + '/' + group.getName());
            }
            WritableMap groupInfo = Arguments.createMap();
            groupInfo.putString("name", group.getName());
            groupInfo.putInt("count", group.getCount());
            groupInfo.putArray("childGroups", childGroupArr);
            groupInfo.putString("path", path + '/' + group.getName());
            groupInfo.putString("type", type);

            groupArr.pushMap(groupInfo);
        }
        return groupArr;
    }

    public static SymbolGroup findSymbolGroups(Resources resources, String type, String path) {

        if (type.equals("")) return null;

        SymbolGroup group;
        String[] pathParams = path.split("/");

        if (type.equals("fill")) {
            group = resources.getFillLibrary().getRootGroup();
        } else if (type.equals("line")) {
            group = resources.getLineLibrary().getRootGroup();
        } else {
            group = resources.getMarkerLibrary().getRootGroup();
        }

        if (pathParams.length > 1) {
            for (int i = 1; i < pathParams.length; i++) {
                group = group.getChildGroups().get(pathParams[i]);
            }
        }
        return group;
    }

    public static WritableArray findSymbolsByGroups(Resources resources, String type, String path) {

        if (type.equals("")) return null;

        WritableArray symbols = Arguments.createArray();
        SymbolGroup group = findSymbolGroups(resources, type, path);
        if (path.equals("")) {
            findSymbolsInGroup(symbols, group, type, path);
        } else {
            for (int i = 0; i< group.getCount(); i++) {
                Symbol symbol = group.get(i);
                WritableMap symbolInfo = Arguments.createMap();

                symbolInfo.putString("groupPath", path);
                symbolInfo.putString("name", symbol.getName());
                symbolInfo.putInt("id", symbol.getID());
                symbolInfo.putString("type", type);

                symbols.pushMap(symbolInfo);
            }
        }
        return symbols;
    }

    public static void findSymbolsInGroup(WritableArray symbols, SymbolGroup group, String type, String path) {
        for (int i = 0; i < group.getCount(); i++) {
            Symbol symbol = group.get(i);
            WritableMap symbolInfo = Arguments.createMap();

            symbolInfo.putString("groupPath", path);
            symbolInfo.putString("name", symbol.getName());
            symbolInfo.putInt("id", symbol.getID());
            symbolInfo.putString("type", type);

            symbols.pushMap(symbolInfo);
        }
        SymbolGroups groups = group.getChildGroups();
        for (int i = 0; i < groups.getCount(); i++) {
            SymbolGroup symbolGroup = groups.get(i);
            findSymbolsInGroup(symbols, symbolGroup, type, path + '/' + symbolGroup.getName());
        }
    }

//    public static List<Symbol> findSymbolsByIDs(Resources commonData, String type, List IDs) {
//        List<Symbol> symbols = new ArrayList<>();
//
//        if (IDs.isEmpty()) {
//            return symbols;
//        }
//
//        for (int i = 0; i < IDs.size(); i++) {
//            Symbol symbol;
//            Object data = IDs.get(i);
//            int id;
//            if (DataUtil.getDataType(data).equals("Double")) {
//                id = ((Double) data).intValue();
//            } else {
//                id = (Integer) data;
//            }
//            if (type.equals("")) {
//                symbol = findSymbolsByID(commonData, id);
//            } else {
//                symbol = findSymbolsByTypeAndID(commonData, type, id);
//            }
//            if (symbol != null) {
//                symbols.add(symbol);
//            }
//        }
//
//        return symbols;
//    }

    public static List<Symbol> findSymbolsByIDs(Resources resources, ReadableArray symbolObjs) {
        List<Symbol> symbols = new ArrayList<>();

        if (symbolObjs.size() == 0) {
            return symbols;
        }

        for (int i = 0; i < symbolObjs.size(); i++) {
            Symbol symbol;
            ReadableMap map = symbolObjs.getMap(i);
            int id = map.getInt("id");
            String type = map.getString("type");
            if (type.equals("")) {
                symbol = findSymbolsByID(resources, id);
            } else {
                symbol = findSymbolsByTypeAndID(resources, type, id);
            }
            if (symbol != null) {
                symbols.add(symbol);
            }
        }

        return symbols;
    }

    public static Symbol findSymbolsByID(Resources resources, int ID) {
        try {
            Symbol symbol;
            symbol = resources.getFillLibrary().findSymbol(ID);
            if (symbol != null) return symbol;

            symbol = resources.getLineLibrary().findSymbol(ID);
            if (symbol != null) return symbol;

            symbol = resources.getMarkerLibrary().findSymbol(ID);
            if (symbol != null) return symbol;

            return symbol;
        } catch (Exception e) {
            return null;
        }
    }

    public static Symbol findSymbolsByTypeAndID(Resources resources, String type, int ID) {
        try {
            Symbol symbol;
            SymbolLibrary lib;
            if (type.equals("fill")) {
                lib = resources.getFillLibrary();
            } else if (type.equals("line")) {
                lib = resources.getLineLibrary();
            } else {
                lib = resources.getMarkerLibrary();
            }
            symbol = lib.findSymbol(ID);

            return symbol;
        } catch (Exception e) {
            return null;
        }
    }
}

File: ios/native/SMSymbol.m
//
//  SMSymbolLibrary.m
//  Supermap
//
//  Created by Yang Shang Long on 2018/11/13.
//  Copyright © 2018 Facebook. All rights reserved.
//

#import "SMSymbol.h"

@implementation SMSymbol
+ (NSArray *)getSymbolGroups:(Resources *)resources type:(NSString *)type path:(NSString *)path {
    SymbolLibrary* symbolLib;
    NSMutableArray* groupArr = [[NSMutableArray alloc] init];
    SymbolGroup* group;
    
    if ([type isEqualToString:@""] && [path isEqualToString:@""]) {
        [groupArr addObject:@{
                              @"name": @"点符号库",
                              @"count": [NSNumber numberWithInteger:resources.lineLibrary.rootGroup.count],
                              @"childGroups": [self findAllSymbolGroups:resources.markerLibrary.rootGroup type:@"marker" path:resources.markerLibrary.rootGroup.name],
                              @"path": resources.markerLibrary.rootGroup.name,
                              @"type": @"marker",
                              }];
        [groupArr addObject:@{
                              @"name": @"线型符号库",
                              @"count": [NSNumber numberWithInteger:resources.lineLibrary.rootGroup.count],
                              @"childGroups": [self findAllSymbolGroups:resources.lineLibrary.rootGroup type:@"line" path:resources.lineLibrary.rootGroup.name],
                              @"path": resources.lineLibrary.rootGroup.name,
                              @"type": @"line",
                              }];
        [groupArr addObject:@{
                              @"name": @"填充符号库",
                              @"count": [NSNumber numberWithInteger:resources.fillLibrary.rootGroup.count],
                              @"childGroups": [self findAllSymbolGroups:resources.fillLibrary.rootGroup type:@"fill" path:resources.fillLibrary.rootGroup.name],
                              @"path": resources.fillLibrary.rootGroup.name,
                              @"type": @"fill",
                              }];
        NSLog(@"%@", resources.markerLibrary.rootGroup.name);
    } else if ([type isEqualToString:@""] && ![path isEqualToString:@""]) {
        [groupArr addObject:[self findAllSymbolGroups:resources.markerLibrary.rootGroup type:@"marker" path:path]];
        [groupArr addObject:[self findAllSymbolGroups:resources.lineLibrary.rootGroup type:@"line" path:path]];
        [groupArr addObject:[self findAllSymbolGroups:resources.fillLibrary.rootGroup type:@"fill" path:path]];
    } else if (![type isEqualToString:@""]) {
        if ([type isEqualToString:@"fill"]) {
            symbolLib = resources.fillLibrary;
        } else if ([type isEqualToString:@"line"]) {
            symbolLib = resources.lineLibrary;
        } else {
            symbolLib = resources.markerLibrary;
        }
        NSString* _path;
        if ([path isEqualToString:@""]) {
            group = symbolLib.rootGroup;
            _path = group.name;
        } else {
            group = [self findSymbolGroups:resources type:type path:path];
            _path = path;
        }
        [groupArr addObject:@{
                              @"name": group.name,
                              @"count": [NSNumber numberWithInteger:group.count],
                              @"childGroups": [self findAllSymbolGroups:group type:type path:_path],
                              @"path": _path,
                              @"type": type,
                              }];
    }
    
    return groupArr;
}

+ (NSMutableArray *)findAllSymbolGroups:(SymbolGroup *)symbolGroup type:(NSString *)type path:(NSString *)path {
    NSMutableArray* groupArr = [[NSMutableArray alloc] init];
    for (int i = 0; i < symbolGroup.childSymbolGroups.count; i++) {
        SymbolGroup* group = [symbolGroup.childSymbolGroups getGroupWithIndex:i];
        NSMutableArray* childGroupArr = [[NSMutableArray alloc] init];
        if (group.childSymbolGroups.count > 0) {
            childGroupArr = [self findAllSymbolGroups:group type:type path:[NSString stringWithFormat:@"%@/%@", path, group.name]];
        }
        [groupArr addObject:@{
                              @"name": group.name,
                              @"count": [NSNumber numberWithInteger:group.count],
                              @"childGroups": childGroupArr,
                              @"path": [NSString stringWithFormat:@"%@/%@", path, group.name],
                              @"type": type,
                              }];
    }
    
    return groupArr;
}

+ (SymbolGroup *)findSymbolGroups:(Resources *)resources type:(NSString *)type path:(NSString *)path {
    if ([type isEqualToString:@""]) return nil;
    
    SymbolGroup* group;
    NSArray* pathParams = [path componentsSeparatedByString:@"/"];
    
    if ([type isEqualToString:@"fill"]) {
        group = resources.fillLibrary.rootGroup;
    } else if ([type isEqualToString:@"line"]) {
        group = resources.lineLibrary.rootGroup;
    } else {
        group = resources.markerLibrary.rootGroup;
    }
    
    if (pathParams.count > 1) {
        for (int i = 1; i < pathParams.count; i++) {
            group = [group.childSymbolGroups getGroupWithName:pathParams[i]];
        }
    }
    
    return group;
}

+ (NSArray *)findSymbolsByGroups:(Resources *)resources type:(NSString *)type path:(NSString *)path {
    if ([type isEqualToString:@""]) return nil;
    NSMutableArray* symbols = [[NSMutableArray alloc] init];
    SymbolGroup* group = [self findSymbolGroups:resources type:type path:path];
    if ([path isEqualToString:@""]) {
        [self findSymbolsInGroup:symbols group:group type:type path:path];
    }
    else
    {
        for (int i = 0; i < group.count; i++) {
            Symbol* symbol = [group getSymbolWithIndex:i];
            NSMutableDictionary *symbolInfo = [[NSMutableDictionary alloc] init];
            
            [symbolInfo setObject:path forKey:@"groupPath"];
            [symbolInfo setObject:[symbol getName] forKey:@"name"];
            [symbolInfo setObject:[NSNumber numberWithInteger:[symbol getID]] forKey:@"id"];
            [symbolInfo setObject:type forKey:@"type"];
            
            [symbols addObject:symbolInfo];
        }
    }
    return symbols;
}

+ (void )findSymbolsInGroup:(NSMutableArray*)symbols group:(SymbolGroup *)group type:(NSString *)type path:(NSString *)path {
    for (int i = 0; i < group.count; i++) {
        Symbol* symbol = [group getSymbolWithIndex:i];
        NSMutableDictionary *symbolInfo = [[NSMutableDictionary alloc] init];
        
        [symbolInfo setObject:path forKey:@"groupPath"];
        [symbolInfo setObject:[symbol getName] forKey:@"name"];
        [symbolInfo setObject:[NSNumber numberWithInteger:[symbol getID]] forKey:@"id"];
        [symbolInfo setObject:type forKey:@"type"];
        
        [symbols addObject:symbolInfo];
    }
    SymbolGroups* groups = group.childSymbolGroups;
    for (int i = 0; i < groups.count; i++) {
        SymbolGroup* symbolGroup = [groups getGroupWithIndex:i];
        [self findSymbolsInGroup:symbols group:symbolGroup type:type path:path];
    }
}

//+ (NSArray *)findSymbolsByIDs:(Resources *)resources type:(NSString *)type IDs:(NSArray *)IDs {
//    if (IDs == nil || IDs.count == 0) return nil;
//    NSMutableArray* symbols = [[NSMutableArray alloc] init];
//    
//    SymbolLibrary* lib;
//    if ([type isEqualToString:@"fill"]) {
//        lib = resources.fillLibrary;
//    } else if ([type isEqualToString:@"line"]) {
//        lib = resources.lineLibrary;
//    } else if ([type isEqualToString:@"marker"]) {
//        lib = resources.markerLibrary;
//    }
//    
//    for (int i = 0; i < IDs.count; i++) {
//        Symbol* symbol;
//        if (type == nil) {
//            symbol = [self findSymbolsByID:resources ID:[(NSNumber *)IDs[i] intValue]];
//        } else {
//            symbol = [self findSymbolsByTypeAndID:resources type:type ID:[(NSNumber *)IDs[i] intValue]];
//        }
//        if (symbol != nil) {
//            [symbols addObject:symbol];
//        }
//    }
//    
//    return symbols;
//}

+ (NSArray *)findSymbolsByIDs:(Resources *)resources symbolObjs:(NSArray *)symbolObjs {
    NSMutableArray* symbols = [[NSMutableArray alloc] init];
    if (symbolObjs == nil || symbolObjs.count == 0) return symbols;
    
    for (int i = 0; i < symbolObjs.count; i++) {
        Symbol* symbol;
        NSDictionary* dic = [symbolObjs objectAtIndex:i];
        NSObject* _id = [dic objectForKey:@"id"];
        NSString* type = [dic objectForKey:@"type"];
        if (type == nil) {
            symbol = [self findSymbolsByID:resources ID:[(NSNumber *)_id intValue]];
        } else {
            symbol = [self findSymbolsByTypeAndID:resources type:type ID:[(NSNumber *)_id intValue]];
        }
        if (symbol != nil) {
            [symbols addObject:symbol];
        }
    }
    
    return symbols;
}

+ (Symbol *)findSymbolsByID:(Resources *)resources ID:(int)ID {
    @try {
        Symbol* symbol;
        symbol = [resources.fillLibrary findSymbolWithID:ID];
        if (symbol) return symbol;
        
        symbol = [resources.lineLibrary findSymbolWithID:ID];
        if (symbol) return symbol;
        
        symbol = [resources.markerLibrary findSymbolWithID:ID];
        if (symbol) return symbol;
        
        return nil;
    } @catch (NSException *exception) {
        return nil;
    }
}

+ (Symbol *)findSymbolsByTypeAndID:(Resources *)resources type:(NSString *)type ID:(int)ID {
    @try {
        Symbol* symbol;
        SymbolLibrary* lib;
        if ([type isEqualToString:@"fill"]) {
            lib = resources.fillLibrary;
        } else if ([type isEqualToString:@"line"]) {
            lib = resources.lineLibrary;
        } else {
            lib = resources.markerLibrary;
        }
        
        symbol = [lib findSymbolWithID:ID];
        if (symbol) return symbol;
        
        return nil;
    } @catch (NSException *exception) {
        return nil;
    }
}


@end
